contracts/asset_manager/src/contract.rs

```diff
@@ -71,48 +71,32 @@ impl AssetManager {
     ) -> Result<(), ContractError> {
         let admin = read_administrator(&env);
         admin.require_auth();
-        let tokens = read_tokens(&env);
-        if tokens.contains(&token_address) {
-            return Err(ContractError::TokenExists);
-        } else {
-            write_tokens(&env, token_address.clone());
-        }
-
         if percentage > POINTS as u32 {
             return Err(ContractError::PercentageShouldBeLessThanOrEqualToPOINTS);
         }
-
-        write_token_data(
-            &env,
-            token_address,
-            TokenData {
-                period,
-                percentage,
-                last_update: env.ledger().timestamp(),
-                current_limit: 0,
-            },
-        );
+        states::write_period(&env, token_address.clone(), period);
+        states::write_percentage(&env, token_address.clone(), percentage);
+        states::write_last_update(&env, token_address.clone(), env.ledger().timestamp());
+        states::write_current_limit(&env, token_address, 0);
         Ok(())
     }
 
     pub fn get_rate_limit(env: Env, token_address: Address) -> Result<(u64, u32, u64, u64), ContractError> {
-        let data: TokenData = read_token_data(&env, token_address)?;
-        
         Ok((
-            data.period,
-            data.percentage,
-            data.last_update,
-            data.current_limit,
+            states::read_period(&env, token_address.clone()),
+            states::read_percentage(&env, token_address.clone()),
+            states::read_last_update(&env, token_address.clone()),
+            states::read_current_limit(&env, token_address),
         ))
     }
 
```

Comment:
same here . its so unintuitive to return tuple of primitive types. Lets wrap the values in struct before returning so that caller knows which value is what. or it should only return current_limit.

Reply:
made change!